packages/mds-compliance/api.ts

```diff
@@ -145,23 +145,55 @@ function api(app: express.Express): express.Express {
     }
     /* istanbul ignore next */
     function fail(err: Error): void {
-      log.error('readDeviceIds fail', err.stack || err).then(() => {
+      log.error(err.stack || err).then(() => {
         res.status(500).send('server error')
       })
     }
 
-    const start_date = now() - days(365)
-    const end_date = now() + days(365)
+    let start_date = now() - days(365)
     const { policy_uuid } = req.params
+    const { provider_id } = req.query
+    let { end_date } = req.query
+
+    if (!end_date) {
+      end_date = now() + days(365)
+    }
 
     if (!isUUID(policy_uuid)) {
       res.status(400).send({ err: 'bad_param' })
+    } else if (end_date) {
+      end_date = parseInt(end_date)
+      start_date = end_date - days(365)
+      try {
+        const policies = await db.readPolicies({ policy_id: policy_uuid, start_date, end_date })
+        const geographies = await db.readGeographies()
+        const deviceIdsWithProvider = await db.readDeviceIds(provider_id)
+        const deviceIds = deviceIdsWithProvider.reduce((acc: UUID[], deviceId) => {
+          acc.push(deviceId.device_id)
+          return acc
+        }, [])
+        const devices = await cache.readDevices(deviceIds)
+        /* eslint-disable-next-line */
+        const deviceMap = devices.reduce((map: {[d: string]: Device}, device) => ((map[device.device_id] = device), map), {})
+        const events = await db.readHistoricalEvents({ provider_id, end_date })
+        const filtered_policies: Policy[] = compliance_engine.filterPolicies(policies)
+        const results: (ComplianceResponse | undefined)[] = filtered_policies.map((policy: Policy) =>
+          compliance_engine.processPolicy(policy, events, geographies, deviceMap)
+        )
+        if (results.length === 0) {
```

Comment:
wat do if all `[undefined, undefined ...]`

Reply:
This should probably be a check for `if (results[0])` instead. In this particular case, we're never going to have `policies` be greater than a length of 1 (`policy_id` is provided as a param to the endpoint), so we'll never have more than one item in the `results` list. So the case `[undefined, undefined, undefined ...]` is impossible here.

Reply:
I'll approve once that change to the results check comes in.